src/gui/editors/AutomationEditor.cpp

```diff
@@ -492,162 +400,228 @@ void AutomationEditor::drawLine( int x0In, float y0, int x1In, float y1 )
+		);
 
-			// will be our iterator in the following loop
-			timeMap::iterator it = time_map.begin();
+		timeMap::iterator clickedNode = getNodeAt(mouseEvent->x(), mouseEvent->y(), editingOutValue);
 
-			// loop through whole time-map...
-			while( it != time_map.end() )
+		switch (m_editMode)
+		{
+			case DRAW:
 			{
-				// and check whether the user clicked on an
-				// existing value
-				if( pos_ticks >= it.key() &&
-					( it+1==time_map.end() ||
-						pos_ticks <= (it+1).key() ) &&
-		( pos_ticks<= it.key() + TimePos::ticksPerBar() *4 / m_ppb ) &&
-		( level == it.value() || mouseEvent->button() == Qt::RightButton ) )
+				m_pattern->addJournalCheckPoint();
+
+				if (m_mouseDownLeft)
 				{
-					break;
-				}
-				++it;
-			}
+					// If we are pressing shift, make a line of nodes from the last
+					// clicked position to this position
+					if (mouseEvent->modifiers() & Qt::ShiftModifier)
+					{
+						drawLine(m_drawLastTick, m_drawLastLevel,
+							posTicks, level);
 
-			if (mouseEvent->button() == Qt::LeftButton)
-			{
-				m_mouseDownLeft = true;
-			}
-			if( mouseEvent->button() == Qt::RightButton )
-			{
-				m_mouseDownRight = true;
-			}
+						// Update the last clicked position for the next time
+						m_drawLastTick = posTicks;
+						m_drawLastLevel = level;
 
-			// left button??
-			if( mouseEvent->button() == Qt::LeftButton &&
-							m_editMode == DRAW )
-			{
-				m_pattern->addJournalCheckPoint();
-				// Connect the dots
-				if( mouseEvent->modifiers() & Qt::ShiftModifier )
-				{
-					drawLine( m_drawLastTick,
-							m_drawLastLevel,
-							pos_ticks, level );
-					m_mouseDownLeft = false;
-				}
-				m_drawLastTick = pos_ticks;
-				m_drawLastLevel = level;
+						// Changes the action to drawing a line of nodes
+						m_action = DRAW_LINE;
+					}
+					else // No shift, we are just creating/moving nodes
+					{
+						// If the place we clicked had no nodes
+						if (clickedNode == tm.end())
+						{
+							// We create a new node and start dragging it
+							TimePos valuePos(posTicks);
+
+							// Starts actually moving/draging the node
+							TimePos newTime = m_pattern->setDragValue(
+								valuePos,
+								level,
+								true,
+								mouseEvent->modifiers() & Qt::ControlModifier
+							);
+
+							// Set the iterator to our newly created node so we can use it later
+							// for the m_moveXOffset calculation
+							clickedNode = tm.find(newTime);
+						}
+						else // If we clicked over an existing node
+						{
+							// Simply start moving/draging it
+							TimePos newTime = m_pattern->setDragValue(
+								TimePos(POS(clickedNode)),
+								level,
+								true,
+								mouseEvent->modifiers() & Qt::ControlModifier
+							);
+
+							// We need to update our iterator because setDragValue removes the node that
+							// is being dragged, so if we don't update it we have a bogus iterator
+							clickedNode = tm.find(newTime);
+						}
```

Comment:
The only difference between these blocks is the `TimePos` value. We can use a ternary operator here instead I think?
```suggestion
						// Start moving/dragging the node
						TimePos newTime = m_pattern->setDragValue(
							// The TimePos of either the clicked node or a new node
							TimePos(
								clickedNode == tm.end()
								? posTicks
								: POS(clickedNode)
							),
							level,
							true,
							mouseEvent->modifiers() & Qt::ControlModifier
						);

						// Set the iterator to our newly created node so we can use it later
						// for the m_moveXOffset calculation
						clickedNode = tm.find(newTime);
```